Introduce the west repository to the Zephyr Project

carlescufi opened the pull request:
This PR introduces the west repository and multi-repository tool to the Zephyr Project.

The changes in this patchset include:

* Removing the built-in copy inside `scripts/` and replace it with the external west repository
* Locating the west copy by using west itself in order to build documentation
* Documenting the new mechanisms to clone west and work with the repositories

Fixes #6205

carlescufi committed:
scripts: remove west from scripts/

west will now be installed via pip in order for the bootstrapper to be
decoupled from the west runners.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
cmake: Use west installation directly

Since we install west from pip and it becomes an executable, use it
directly assuming it's in the user's PATH.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>
Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

carlescufi committed:
doc: scripts: Retrieve west path at runtime

Since west is no longer part of the Zephyr tree, we can no longer
hardcode its path anymore. Instead, use west itself to retrieve its
path, in order to point the documentation build to the correct
active west installation.
This is optional, and the documentation build will still work if west is
not installed on the system.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
doc: Use west instead of Git to clone Zephyr

Since west is no longer included in the Zephyr repository, instruct
users to install and use west in order to obtain the Zephyr source code.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>
Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

nashif committed:
ci: Use west to clone zephyr

West right now modifies the zephyr tree and points to master. In CI we
need to point to the pull-request HEAD instead, so so whatever shippable
does after cloning with west.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

dbkinder commented on doc/getting_started/getting_started.rst:
```diff
@@ -20,22 +20,69 @@ Follow one of the following guides for your host operating system.
+   west clone
+
+.. note::
+   You can replace :file:`zephyrproject` with the folder name of your choice.
```
Add:
```
West will create the named folder if it doesn't already exist.
If no folder name is specified, west initializes the current
working directory.
```

mbolivar commented on doc/conf.py:
```diff
@@ -26,11 +27,20 @@
+    # Add west, to be able to pull in its API docs.
+    sys.path.append(os.path.join(west_path, 'src'))
+except FileNotFoundError as e:
+    raise RuntimeError('Unable to find west. Make sure you have installed'
```
missing spaces at end of each string. This would read as 'installedit' and the next lines would result in 'installationusing'.

dbkinder commented on doc/west/index.rst:
```diff
@@ -1,30 +1,139 @@
+``west init`` will do the following:
+
+* Clone west itself in a :file:`west/west` folder
+* Clone the manifest in a :file:`west/manifest` folder (additional information
```
```suggestion
* Clone the manifest in a :file:`manifest` folder (additional information
```

mbolivar reviewed (commented):
Just one nit I didn't notice in the previous review.

dbkinder commented on doc/west/index.rst:
```diff
@@ -1,30 +1,139 @@
+Installation
+============
+
+A west installation is the result of running ``west init`` on a folder. As
```
```suggestion
A west installation is the result of running ``west init`` in an existing folder,
or ``west init <foldername>`` to create and initialize a folder . As
```

dbkinder commented on doc/west/index.rst:
```diff
@@ -1,30 +1,139 @@
+
+.. code-block:: console
+
+   ├── west
```
to avoid confusion (and mirror the example in the getting started guide) lets call this top level folder ``zephyrproject``
```suggestion
   ├── zephyrproject
```

dbkinder commented on doc/west/index.rst:
```diff
@@ -1,30 +1,139 @@
+   │   └── west
+   └── zephyr
+
+Additionally, ``west init`` creates a configuration file :file:`west/config`
```
to be clear, is this file ``zephyrproject/west/config`` or ``zephyrproject/config``?  (your picture above had two folders named ``west``)

carlescufi replied:
this is `zephyrproject/west/config`

carlescufi replied:
I've tried to clarify this in my diagram: `west init` creates a `west` folder with 3 items:

* A `west` subfolder
* A `manifest` subfolder
* A `config` file

dbkinder replied:
Ah, I see now. (I just tried doing a ``west init``).

dbkinder commented on doc/west/index.rst:
```diff
@@ -1,30 +1,139 @@
+:ref:`west-multi-repo`.
+
+The west bootstrapper delegates handling for all commands except ``init`` to
+the installation in :file:`west/west`.
```
```suggestion
the installation in :file:`zephyrproject/west`.
```

carlescufi replied:
See new diagram once I push

dbkinder commented on doc/west/repo-tool.rst:
```diff
@@ -217,8 +217,8 @@ to the latest manifest revision, and inspects upstream changes.
 
-      $ west init
-      $ west clone
+      west init
```
Because ``west init`` will initialize the **current** folder, it would be good to add a preamble to the workflow to explain setting up the development working directory.  For example, should a developer start with an empty folder (e.g., ``~/zephyrproject``) and cd into it before doing this ``west init`` command?

carlescufi replied:
I've added something like that, mirroring the instructions in Getting Started.

dbkinder commented on doc/release-notes.rst:
```diff
@@ -9,17 +9,16 @@ Zephyr project is provided as source code and build scripts for different
 
+With the introduction of the :ref:`west` tool after the release of Zephyr 1.13,
+it is no longer recommended to download or clone the source code manually.
+Instead we recommend to follow the instructions in :ref:`get_the_code` to do
```
Also, we'll need to update the instructions at the bottom of https://www.zephyrproject.org/developers/#releases-overview.

Will subsystems/networking/qemu_setup.rst need updating too?  (It talks about git cloning the zephyrproject-rtos/net-tools repo)

carlescufi replied:
I might be missing something, I cannot find the text you mention.

dbkinder replied:
The bottom of the releases overview says, "All Zephyr project source code is maintained in a Git repository; you can either download source as a tar.gz file (see the bottom of the GitHub release notes pages listed below) or use Git."  Sounds like we should probably just delete this and refer to the release notes page in GitHub here.

carlescufi commented:
@dbkinder @mbolivar comments addressed

dbkinder reviewed (approved):
LGTM, thanks!

aescolar commented:
CC @fnde-ot

carlescufi commented:
recheck

aescolar reviewed (changes requested):
I'm quite concerned about the direction this it taking.
(Mind you, I can understand the perspective of the `west` developers. You have some understanding about what users want, and are trying to make it dumb easy to do that. That is very laudable.
My concern stems from the fact that it seems you are in the process of making it way harder to do what you do not assume people wants to do.)

The thing is, for my company, Zephyr is just one component in a much bigger project, and `west` does not provide anything we need or want.
Yet west design seems to assume a lot about what others are trying to do.
It is clear that for us it will just create extra work and problems.

Therefore seeing `west` being mandated and cemented all around, together with some of the design choices, makes me wonder how much manpower we will be forced to use working around it.

------

Apart from this mayor issue, a few other more minor review comments.

tejlmand commented on doc/west/index.rst:
```diff
@@ -54,8 +163,9 @@ Design Constraints
-  command-line tools, i.e. use Zephyr without using West. It may not
-  always be *convenient* to do so, however. (If all of West's features
+  command-line tools, i.e. use Zephyr without using west (although west itself
+  might need to be installed and accessible to the build system). It may not
```
meaning, the tools inside Zephyr, that was already `west`, such as `flashing, debug, attach, ...`

Those are now in `west` repo, so in order to have those working, you would need the `west` repo.
But if not using `west` for flashing, debug, etc. I assume you can leave it out, although we might need to ensure CMake can handle it, in case `west` is not present.

aescolar replied:
> I assume you can leave it out

That being the point

carlescufi replied:
It depends what you do really. If you are not going to flash or debug, then you definitely not need it. In the future however, you will need it in order to build, since the build system will use `west` to locate the different repositories available for building.

tejlmand replied:
question is, if people really want / has a reason to avoid cloning `west`.
And Zephyr needs it for tje build system to locate all required repositories for building, then could we provide an easy to do so way, e.g. arguments to CMake. to allow people to do so.

Simply, not preventing them from doing so.
That is give the option that what `west` returns can simply be provided manually ?

aescolar replied:
> In the future however, you will need it in order to build

Ok.. so today you can still build images with that warning, but not doc. And soon nothing. Meaning, not optional :-p

carlescufi commented on .shippable.yml:
```diff
@@ -32,6 +32,14 @@ build:
         options: "-e HOME=/home/buildslave --privileged=true --tty --net=bridge --user buildslave"
 
     ci:
+      - sudo pip3 install west
```
We already depend on many other python modules that are not fixed to a particular version. The latest revision of `west` is guaranteed to work with the latest master. If we need to pin a particular release to a version of `west` this can be done in the repository itself

aescolar replied:
> that are not fixed to a particular version. 

In CI they are. You know that 2 CI runs using the same docker image will use the same version of that tool.
With this change, if somebody pushes to west's master a breaking change (say a bug), things will suddenly break here. That simply should not be a possibility in a proper CI environment. Here we are testing a Zephyr PR, therefore everything else should be constant.

nashif replied:
west is already in the CI docker image, so this can be removed.

carlescufi commented on cmake/host-tools.cmake:
```diff
@@ -1,5 +1,17 @@
+  )
+if(${WEST} STREQUAL WEST-NOTFOUND)
+  unset(WEST)
+  message(WARNING "Unable to find west, some functionality may not be available. \
```
Well, I would argue that if the user is not going to be able to do things like `ninja flash` and `ninja debug` we should warn them early.

aescolar replied:
What if the user (like all developers in my company) does not want to do that?

carlescufi commented on doc/getting_started/getting_started.rst:
```diff
@@ -20,22 +20,73 @@ Follow one of the following guides for your host operating system.
+   from this empty enclosing folder :file:`zephyrproject/`
+   run ``west init`` and then ``west clone``.
+
+This will create a full Zephyr installation, cloning west itself along with any
```
I chose this word because `west init` does more than just fetching Zephyr. It actually installs `west` itself in `<folder>/west/west` alongside the manifest. So in essence you are installing the `west` tool (`pip` only installs the bootstrapper) and then fetching all the repositories required (right now only one, but multiple in the future).

aescolar replied:
> It actually installs west itself  in `<folder>/west/west`

So it clones West and Zephyr into that folder? My point is that here you don't "install" anything in the way people understands "installing". (while pip does install things in the way people understands installing)

nashif replied:
agree with @aescolar here, this is not a "full Zephyr installation", maybe 'zephyr development environment setup"?

carlescufi commented on doc/conf.py:
```diff
@@ -26,11 +27,20 @@
+    west_path = desc.decode(sys.getdefaultencoding()).strip()
+    # Add west, to be able to pull in its API docs.
+    sys.path.append(os.path.join(west_path, 'src'))
+except FileNotFoundError as e:
```
We are only using `west` to locate its own installation in order to document its API.

aescolar replied:
Good to know. But this exception handling will break making the doc if west is not found.

carlescufi commented on doc/release-notes.rst:
```diff
@@ -9,17 +9,16 @@ Zephyr project is provided as source code and build scripts for different
+release), or clone the GitHub repository.
 
+With the introduction of the :ref:`west` tool after the release of Zephyr 1.13,
+it is no longer recommended to download or clone the source code manually.
```
recommended. It definitely is possible, by cloning west itself and then the repositories manually.

tejlmand replied:
should still be possible to do manually.
but tools located in `west` will of course not be available.

carlescufi commented on doc/west/index.rst:
```diff
@@ -1,30 +1,142 @@
+
+West is downloaded and installed on a system in two stages:
+
+* Bootstrapper: Installed using ``pip``, implements ``west init``
```
I will let @mbolivar answer this one since he designed the 2-step installation process in the first place, which mirrors Google repo's.

aescolar replied:
>  which mirrors Google repo's.

Android's repo just needs the repo script to be somewhere to be executed. Not to be installed with `pip`. See for example:
https://github.com/aescolar/ci-dockerfiles/blob/master/shippable/install_bsim.sh#L6 
If I can do that with west, without sideeffects somewhere else in the host, then I'm happy.

tejlmand replied:
RFC from where we can discuss changes to allow `west` bootstrapper to be fetched without `pip`
https://github.com/zephyrproject-rtos/west/pull/133

tejlmand commented:
> My concern stems from the fact that it seems you are in the process of making it way harder to do what you do not assume people wants to do

And that is a very valid concern.
`west` should be an aid, and although splitting Zephyr into multi-repo makes it harder to get all the right repo, when doing things by hand / other tools.
But `west` should not make it impossible to do things in other ways.

If this is the case, we should address it.

tejlmand commented on doc/conf.py:
```diff
@@ -26,11 +27,20 @@
+    sys.path.append(os.path.join(west_path, 'src'))
+except FileNotFoundError as e:
+    raise RuntimeError('Unable to find west. Make sure you have installed '
+                       'it using pip and used it to create an installation '
```
See your point.

We tried to keep a consistent way to refer to the `Zephyr installation`.
Will try to change phrasing.

tejlmand commented on doc/west/index.rst:
```diff
@@ -1,30 +1,142 @@
+       └── zephyr
+
+Additionally, ``west init`` creates a configuration file :file:`west/config`
+that stores configuration metadata about the west installation.
```
no, this config file is about `west` and `manifest` repo themselves.

The Zephyr project is described in the manifest file.

aescolar commented:
> split up the zephyr repository into a "core" 

That is perfectly fine. If all west did for multi repo was the same that, say, Android's repo does we would be fine. That being because all android repo does, can be redone by hand or by another script trivially.
(It just fetches you several repos as listed in a manifest, and in the worst case it creates a couple of symlinks. And, if you want, you can use it to run commands in several of those repos simultaneously)

> you can still use it and then run west init after cloning.

(rhetorical question) If I have already cloned everything, why?

> But west should not make it impossible to do things in other ways.

That is exactly what I hope for.
I think the problem stems from aiming at integrating things that, IMHO, don't belong together.
* One thing is fetching repositories
* Another is building code

If there is any need for interfaces between these steps, those interfaces should be very clean and very human accessible. So that in any of those steps you could use other tools.
So, what information do we need to pass to the build system? A list of repos and filesystem paths? Would a plain text file do the job? If so, then have multirepo-west generate that text file and cmake read it. Let users modify or create that text file by hand or with other tools if they need.

> both Zephyr's and each member company's goals using a common toolset.

I think this is the wrong overall approach. It makes sense to provide a default & easy way of doing things. But you should not aim at encompassing everybody's needs with the same tool all the steps of the way. Otherwise you end up with massive complexities to accommodate doing trivialities in many different ways.
Having clear interfaces, between clearly separate steps performed by clear one purpose tools: This allows most people to use those common one purpose tools when they fit their needs, and something else when they don't.
Attempting a common "god" tool that knows it all and does it all, is just aiming way too high, and ensuring that anybody who deviates a bit of the typical path won't be able to use any of it.

ulfalizer commented:
> That is exactly what I hope for.
> I think the problem stems from aiming at integrating things that, IMHO, don't belong together.
> 
> * One thing is fetching repositories
> * Another is building code
> 
> If there is any need for interfaces between these steps, those interfaces should be very clean and very human accessible. So that in any of those steps you could use other tools.
> So, what information do we need to pass to the build system? A list of repos and filesystem paths? Would a plain text file do the job? If so, then have multirepo-west generate that text file and cmake read it. Let users modify or create that text file by hand or with other tools if they need.
> 
> > both Zephyr's and each member company's goals using a common toolset.
> 
> I think this is the wrong overall approach. It makes sense to provide a default & easy way of doing things. But you should not aim at encompassing everybody's needs with the same tool all the steps of the way. Otherwise you end up with massive complexities to accommodate doing trivialities in many different ways.
> Having clear interfaces, between clearly separate steps performed by clear one purpose tools: This allows most people to use those common one purpose tools when they fit their needs, and something else when they don't.
> Attempting a common "god" tool that knows it all and does it all, is just aiming way too high, and ensuring that anybody who deviates a bit of the typical path won't be able to use any of it.

I've been arguing for this as well. It feels a bit weird to me to bundle a `repo`-like with building and flashing helpers.

Building and flashing are so closely tied to the Zephyr repository that I think it would make the most sense to implement any helpers there, as a few simple scripts. That way you also know that they will stay in sync with the underlying things they call through to.

Another plus is that working on the Zephyr code is easier than working on the West code, due to the wrapper and self-update mechanisms (which got inherited from `repo`).

I'm not big on "god" tools either. Simple single-purpose tools tend to be easier to understand and work with, and give better modularity. If you try to do too much in one place or force people's hands too much, you often end up with an impenetrable black box.

For the multi-repo stuff, we should be sure that splitting out repositories makes people's workflow smoother. It's one more tool to learn, so other things have to offset it.

carlescufi commented:
@aescolar I have now made west fully optional. It will build the doc without it and, even though it will issue a message during CMake configuration, it is no longer a warning. Also I have removed the `pip3 install west` from the shippable configuration.

@mbolivar @dbkinder Please take a look at the changes I had to make to get the doc to build without west.

dbkinder reviewed (approved):
I'm working on adding improved ``.. only::`` directive handling that would eliminate the need for using the exclude hack in conf.py, but until then, this seems a viable solution.  I'll update the doc processing later to clean this up.

aescolar commented on .shippable.yml:
```diff
@@ -32,6 +32,13 @@ build:
 
     ci:
+      - cd ..
+      - west init .
```
So here you will be fetching whatever the default manifest says?
With this, we do not know what is actually run in CI anymore (it depends on a external repo that changes on its own)
== Not ok.*1

To do this properly, you would need to fetch from here a particular version of a particular manifest.
It should be a commit on its own to change which manifest version is used.

*1 Just think about the case when you will be moving some ext/ component to a external repo. You will on one side change the manifest, while on Zephyr you will do a PR to remove that ext/ component and change how its built. While you have changed the manifest and until the relevant changes have been merged in Zephyr, Zephyr will be broken.

aescolar replied:
Yes today. But once you start using for what it is meant to be used : handling multiple repositories it needs to be handled as I commented.

aescolar commented on cmake/host-tools.cmake:
```diff
@@ -1,5 +1,17 @@
+  )
+if(${WEST} STREQUAL WEST-NOTFOUND)
+  unset(WEST)
+  message(STATUS "Unable to find west, some functionality may not be available. \
```
This is going to be very annoying (cmake is already way too verbose). Please remove this.
Won't the user get a clear error if he tries to run `ninja flash` and west is not found?

ulfalizer replied:
Yeah, I think this is another thing that could be solved by moving the functionality into the Zephyr repository.

West in turn calls back to CMake as far as I can tell. That makes it very twisty and hard to follow.

carlescufi replied:
Fair enough points from both. I will remove it.

tejlmand replied:
> This is going to be very annoying (cmake is already way too verbose). Please remove this.
> Won't the user get a clear error if he tries to run `ninja flash` and west is not found?

CMake now detects whether `west` is available and useable and issue a corresponding message when issueing, e.g. `ninja flash/debug/...`.

Example:
```
West was not found in path. To support '/usr/bin/ninja flash', please install west bootstrapper with: 'pip install west --user', and thereafter 'west init'.
```

aescolar commented on doc/getting_started/getting_started.rst:
```diff
@@ -20,22 +20,73 @@ Follow one of the following guides for your host operating system.
 
-   git clone https://github.com/zephyrproject-rtos/zephyr
+   # Linux
+   pip3 install --user west
```
Is it possible to just wget the west bootstrapper?

tejlmand replied:
> 
> 
> Is it possible to just wget the west bootstrapper?

check this RFC, and comment whether this is the behavior that you also think `west` should support.
https://github.com/zephyrproject-rtos/west/pull/133

mbolivar replied:
> Is it possible to just wget the west bootstrapper?

Yes. We publish wheels and sdists to pypi. You can `wget` the 0.4.0 ones here:

https://files.pythonhosted.org/packages/ba/cf/260bd15d72ca9a9d9bf6006575fb8ef2659dde6334b3f977ca88cc6a9951/west-0.4.0-py3-none-any.whl
https://files.pythonhosted.org/packages/c6/7f/26501ecf53303beac2f8c9729cf9bf25b053d4901674a62b670d91e774a7/west-0.4.0.tar.gz

You can then install them using setuptools or pip or whatever you like.

mbolivar replied:
> There was a push to put everything in the "West library".

s/library/package/ then yes. We discussed ad nauseam that package != library.

> It's still independent at least.

And, by design, always will be.

>> You can then install them using setuptools or pip or whatever you like.
> Yes. We publish wheels and sdists to pypi. You can wget the 0.4.0 ones here:
> [...]
> You can then install them using setuptools or pip or whatever you like.

(@aescolar I'm assuming `python setup.py install` on an sdist to a standalone directory that doesn't affect your system path might be acceptable).

The main reason (IMO) why we can't just distribute a Python script has to do with Windows. We want every operating system to be able to run `west <command>` and get the same results. To do that, our setup.py uses `entry_points` to ask for a `west` to be installed in a system-dependent manner.

On Linux, that looks like this:

```
#!/home/mbolivar/.local/share/virtualenvs/west/bin/python3

# -*- coding: utf-8 -*-
import re
import sys

from west._bootstrap.main import main

if __name__ == '__main__':
    sys.argv[0] = re.sub(r'(-script\.pyw?|\.exe)?$', '', sys.argv[0])
    sys.exit(main())
```

On Windows, though, it's a .exe:

```
>>> import shutil
>> shutil.which('west')
'C:\\Python36\\Scripts\\west.EXE'
```

tejlmand replied:
> > Is it possible to just wget the west bootstrapper?
> Yes. We publish wheels and sdists to pypi. You can wget the 0.4.0 ones here:

@mbolivar I think that what @aescolar is asking for, it the possibility to just `wget` a single file, and continue from there, just as you do with `repo`.
But that's just how I read his comments, hopefully @aescolar can make his thinking clearer to both of us :)

> The main reason (IMO) why we can't just distribute a Python script has to do with Windows. We want every operating system to be able to run west <command> and get the same results. To do that, our setup.py uses entry_points to ask for a west to be installed in a system-dependent manner.

And that we already achieve with current design.
Question is, should we allow for other ways as well ?
I don't see a blocker or additional cost in also providing the possibility to fetch the bootstrapper as a single file, if there is a wish for that.

Then users can choose their prefered way, while we recommend the `pip` installation.

aescolar replied:
Yes, my comment was about wondering if it would be possible to have such a "local" way of getting the bootstrapper as that of repo. If it is not and it could be added, I think it would be nice for some users. But in this PR context, it was just a question => resolving the conversation.

aescolar commented on doc/getting_started/getting_started.rst:
```diff
@@ -20,22 +20,73 @@ Follow one of the following guides for your host operating system.
+
+.. code-block:: console
+
+   west init zephyrproject
```
What is the process for changing the default manifest in the manifest repository? Are there PRs and are those tested thru shippable also?

carlescufi replied:
There are PRs but they are not tested by shippable yet.

aescolar replied:
Then more weight to the argument of not just fetching the master tip here.

nashif replied:
right now west is not pulling the zephyr tree, because this was already fetched by shippable and the following lines deal with pointing git to the code in the PR, so we have exactly the same as we have things today without west. How to deal with multi-repo changes and manifest versions is not subject of this PR AFAIK.

ulfalizer commented on doc/west/index.rst:
```diff
@@ -1,30 +1,139 @@
+first in order to use the rest of functionality included in ``west``, since
+``west init`` will do the following:
+
+* Clone west itself in a :file:`west/west` folder
```
I was arguing for having `.west`, like repo has `.repo` (though you don't get building and flashing helpers in there), but people said that users would be horribly confused by the leading dot.

mbolivar replied:
> people said that users would be horribly confused by the leading dot.

Actually, people said that we should be consistent on Windows and Linux where file names are concerned.

carlescufi commented:
@aescolar 

>So it clones West and Zephyr into that folder? My point is that here you don't "install" anything in the way people understands "installing". (while pip does install things in the way people understands installing)

Yes but it clones a program that you are going to run immediately after by typing `west <command>`. It also "installs" in the sense of downloading and placing a program that you are going to run. I am not married to the term, but this is the most representative one that came to mind.

mped-oticon commented on doc/west/index.rst:
```diff
@@ -1,30 +1,142 @@
+
+West was added to the Zephyr project to fulfill two fundamental requirements:
+
+* The ability to work with multiple Git repositories
```
I know we talked about this already, but for the benefit of everybody;
Why was `repo` not deemed fit for working with multiple Git repositories?

carlescufi replied:
In short:

* Python 2 only
* Uses symlinks so it doesn't work on windows without an admin command prompt
* Hardcoded gerrit code review
* Poorly documented and maintained

marc-hb replied:
I think these repo shortcomings are important to mention somewhere in the documentation itself and not just in this code review, otherwise west-multirepo looks like just another NIH tool.

While adding them drop "poorly documented and maintained" or justify it a bit https://gerrit.googlesource.com/git-repo/
https://gerrit.googlesource.com/git-repo/+log/master

Repo is used by a fair number of projects despite these shortcomings. So you are indirectly saying the world could use a better multirepo tool. I could agree. 
 => Why should [west-]multirepo be specific to Zephyr?

nashif commented:
It would be helpful to have a section in the documentation describing how to use zephyr if you do not want to use west or if you have your own way of managing multiple git repositories .

carlescufi commented:
> It would be helpful to have a section in the documentation describing how to use zephyr if you do not want to use west or if you have your own way of managing multiple git repositories .

OK, I will add this now.

carlescufi commented:
@nashif comments addressed

aescolar commented:
>> It would be helpful to have a section in the documentation describing how to use zephyr if you do not want to use west or if you have your own way of managing multiple git repositories .

> OK, I will add this now.

Did you get to add this one? I would also be happy to see it.

aescolar commented:
>> It would be helpful to have a section in the documentation describing how to use zephyr if you do not want to use west or if you have your own way of managing multiple git repositories .

> OK, I will add this now.

@carlescufi  Did you get to add this one? I would also be happy to see it.

carlescufi commented:
@mbolivar @aescolar @tejlmand I added a short section about sidestepping west here:
https://github.com/zephyrproject-rtos/zephyr/pull/11715/files#diff-3a7071e04a74d8d9bdc5067cf8748c96R198

aescolar commented on doc/west/index.rst:
```diff
@@ -78,11 +188,63 @@ West is:
+   cd west
+   git clone https://github.com/zephyrproject-rtos/west.git
+   git clone https://github.com/zephyrproject-rtos/manifest.git
+   echo > .west_topdir
```
`touch .west_topdir` doesn't do it?

carlescufi commented:
@aescolar `touch` only works on *NIX. `echo >` works both on *NIX and Windows

carlescufi committed:
doc: west: Move files to the correct location

West documentation files now belong in doc/tools/west and not doc/west.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
doc: west: Overhaul documentation to match current model

Overhaul the west documentation so that it matches the model that is
currently implemented in the west repository.
This model is no longer subject to change in the short or mid term,
since it has been selected as the only viable one for multi-repo
management using west.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

Commit:
scripts: add west build, flash, and debug commands

West now supports a mechanism for extension commands. Use it to move
the command implementations that are tightly coupled with boards and
the zephyr build system back into the Zephyr repository.

This patch doesn't include test cases. Those will be moved over in a
subsequent patch.

Signed-off-by: Marti Bolivar <marti@foundries.io>

carlescufi committed:
ci: Update to the new west model

The new west is able to initialize itself from an existing zephyr clone
without modifying it, adapt the CI script accordingly.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

Commit:
scripts: run_ci.sh: remove dead code

Commit 88ece494 ("ci: do not build docs in main CI job") removed the
call to build_docs but not the function itself. That makes it
confusing to try to find where it's called; remove it.

Signed-off-by: Marti Bolivar <marti@foundries.io>

Commit:
scripts: run_ci.sh: add pytest-based testing for west commands

Incorporate these into run_ci.sh the same way that btsim results were
done. This adds a dependency on pytest.

Signed-off-by: Marti Bolivar <marti@foundries.io>

Commit:
west: add west.yml

This is the west manifest file for mainline zephyr.

For a burn-in period, we are going to keep west pointing at master to
avoid deps rolls as we iron out any kinks.  We can then set its
revision to a SHA after things settle down.

The projects key is mandatory (if there's only one repository, there
wouldn't be much need for west!) so we'll start things off with the
net-tools repository at its current master SHA.

Signed-off-by: Marti Bolivar <marti@foundries.io>

Commit:
.shippable.yml: HACK: install latest west

After some discussion, we've decided to ensure that shippable runs
with the latest west bootstrapper on PyPI by making that happen in
.shippable.yml for now, rather than updating our CI container.

Signed-off-by: Marti Bolivar <marti@foundries.io>

carlescufi commented:
recheck

mbolivar commented:
recheck

carlescufi committed:
west: Multiple CI fixes for the topic-west branch

In order to be able to merge the topic branch, we require a few
fixes to CI.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>
Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi commented:
recheck

marc-hb commented on doc/tools/west/index.rst:
```diff
@@ -78,11 +216,61 @@ West is:
+This section applies to Zephyr users who fall into one or more of these
+categories:
+
+- Already use a multi-repository management system (for example Google repo)
```
So there's a tradition for this sort of tool to have un-googlable names...

nashif replied:
LOL

carlescufi replied:
😅 I must admit I had not considered this until you brought it up. I guess you'll need to google "zephyr west". But even then you may get results about the western wind 😮

carlescufi committed:
doc: west: Misc formatting fixes

Miscellaneous formatting fixes in the west documentation.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
doc: west: Do not fail doc build if west not used

In case west is not present or was not used to initialize the local
zephyr repository, avoid failing the documentation build by including a
placeholder file.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
scripts: west: Add support for nRF91 in nrfjprog

Extend the list with the new nRF91 SoC.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>
Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

carlescufi committed:
doc: getting_started: Clarify what west does

Add additional details about what running `west init` and `west update`
actually does in the local filesystem.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

Commit:
west_commands: add information on running tests

Add information to the README on how to run the tests, and note that
they are part of Zephyr's CI.

Signed-off-by: Marti Bolivar <marti@foundries.io>

Commit:
scripts: west commands: add text for "west --help"

Although each of these command implementations has good output for
"west <command> --help", the "west --help" output is missing the
one-line descriptions.

Add them by modifying west-commands.yml to use a new 'help' key in
each command name. These need to be kept in sync with the Python
sources. For now just do that by copy/pasting. We could add fancy
logic to load the help from west-commands.yml later if we want.

Signed-off-by: Marti Bolivar <marti@foundries.io>
